### PR TITLE
Reduce the memory footprint of the training script and few other bug fixes

### LINKED ISSUE #193: Train the model on chinese OntoNotes 5.0 but eat up all my 64GB memory
I am training it for Chinese Model to use for event extraction for my company, but I only use part of English data for testing the model, as show in first image(train data ). But it eats up 50% memory(64GB) when build coreference clusters，when I use all data from conll-2012, my memery is 100% and can't run any more. Can you tell me how much memory should I use to train the model, or any other requirements? The attach file is my training machine performance.
traindata
gpu
computer

### PR DESCRIPTION
**Pull request by Abhijit-2592**
Fixes #193 
This Pull request addresses the following 3 problems:

1. Issue #193 highlights that the training script hogs a huge amount of RAM memory. Looking into the code base, this is mainly attributed to: Loading the entire `.npy` files to memory. Thus I have provided an option to load them **lazily** so that RAM usage does not blow up. This option can be turned on by passing a `--lazy` flag to `learn.py`. Before this I was unable to train on the dataset (approx 8.5GB size in disk) generated using spacy's `en_core_web_lg` model on my laptop (16GB RAM and 6GB GPU). The dataset required more than 50GB of RAM to train. After this change the memory foot print is ~ `5GB` for the same dataset and runs on my laptop without hiccups.
2. While creating the dataset it would be beneficial if user could pass the required `spacy model`.
3. The pull request #230 attempts to fix training on GPU but introduces a new bug which throws an error while training on CPU. This is fixed here. Now the training can be done both on GPU and CPU as required.

### CODE REVIEW

File: neuralcoref/train/conllparser.py [hunk 1]
```diff
@@ -595,20 +595,24 @@ def read_corpus(self, data_path, debug=False):
...
-                    print("Loaded model", model_option)
-                except:
-                    print("Could not detect model", model_option)
-        if not model:
-            print("Could not detect any suitable English model")
-            return
 
+        if model is None:
+            model_options = ['en_core_web_lg', 'en_core_web_md', 'en_core_web_sm', 'en']
+            model = None  # this declaration is redundant
```
Comment (svlandeg) [inline neuralcoref/train/conllparser.py]: if it's redundant (and it is in the current version of the code), feel free to just remove it ;-)

File: neuralcoref/train/conllparser.py [hunk 2]
```diff
@@ -722,6 +726,7 @@ def _vocabulary_to_file(path, vocabulary):
     parser.add_argument('--n_jobs', type=int, default=1, help='Number of parallel jobs (default 1)')
     parser.add_argument('--gold_mentions', type=int, default=0, help='Use gold mentions (1) or not (0, default)')
     parser.add_argument('--blacklist', type=int, default=0, help='Use blacklist (1) or not (0, default)')
+    parser.add_argument('--spacy_model', type=str, default=None, help='spacys language model')
```
Comment (svlandeg) [inline neuralcoref/train/conllparser.py]: `spacys` looks weird to me, I would put `spaCy's` or just simply call this "model name" as in other example scripts in spaCy. Note that this repo will be integrated more closely into spaCy in the future anyway.

File: neuralcoref/train/dataset.py [hunk 3]
```diff
@@ -44,13 +70,18 @@ def __init__(self, data_path, params, no_targets=False):
                 continue
             numpy_files_found = True
             print(file_name, end=', ')
-            datas[file_name.split(u'.')[0]] = np.load(data_path + file_name)
+            datas[file_name.split(u'.')[0]] = np.load(data_path + file_name, mmap_mode="r" if params.lazy else None)
```
Comment (svlandeg) [inline neuralcoref/train/dataset.py]: Nice little fix !

File: neuralcoref/train/learn.py [hunk 4]
```diff
@@ -303,8 +303,8 @@ def run_epochs(start_epoch, end_epoch, loss_func, optim_func, save_name, lr, g_s
     parser.add_argument('--min_lr', type=float, default=2e-8, help='min learning rate')
     parser.add_argument('--on_eval_decrease', type=str, default='nothing',
                         help='What to do when evaluation decreases ("nothing", "divide_lr", "next_stage", "divide_then_next")')
+    parser.add_argument('--lazy', action='store_true', help='Use lazy loading while loading the npy files')
```
Comment (svlandeg) [inline neuralcoref/train/learn.py]: I would define the type, default and make it clear in the description:
`type=int, default=1, help='Use lazy loading (1, default) or not (0) while loading the npy files'`

I think this setting can be on by default :-)

File: neuralcoref/train/dataset.py (hunk 3 above)
Comment (Abhijit-2592) [reply neuralcoref/train/dataset.py]: Thanks

File: neuralcoref/train/conllparser.py (hunk 1 above)
Comment (Abhijit-2592) [reply neuralcoref/train/conllparser.py]: Removed it

File: neuralcoref/train/conllparser.py (hunk 2 above)
Comment (Abhijit-2592) [reply neuralcoref/train/conllparser.py]: Changed to `model name`

File: neuralcoref/train/learn.py (hunk 4 above)
Comment (Abhijit-2592) [reply neuralcoref/train/learn.py]: Done!

**Comment by Abhijit-2592** [comment]
@svlandeg done

**Comment by EricLe-dev** [comment]
It seems did not work. I trained a dataset of approximately 8.5 Gb on a machine with 62GB of RAM and 8GB GPU. Still I got this error:
image

**Comment by svlandeg** [comment]
@EricLe-dev : apologies for the late follow-up, but did you pull the recent `master` branch for neuralcoref and did you install it from source? Because this fix isn't part of any release yet...

**Pull request merged**